examples/twentyseven-pt-stencil-solver/twentyseven-pt-stencil-solver.cpp

```diff
@@ -0,0 +1,365 @@
+
+    // Figure out where to run the code
+    const auto omp = gko::OmpExecutor::create();
+    std::map<std::string, std::shared_ptr<gko::Executor>> exec_map{
+        {"omp", omp},
+        {"cuda", gko::CudaExecutor::create(0, omp)},
+        {"reference", gko::ReferenceExecutor::create()}};
+    // executor where Ginkgo will perform the computation
+    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    // executor where the application initialized the data
+    const auto app_exec = exec_map["omp"];
+
+    // Tell Ginkgo to use the data in our application
+
+    // Matrix: we have to set the executor of the matrix to the one where we
+    // want SpMVs to run (in this case `exec`). When creating array views, we
+    // have to specify the executor where the data is (in this case `app_exec`).
+    //
+    // If the two do not match, Ginkgo will automatically create a copy of the
+    // data on `exec` (however, it will not copy the data back once it is done
+    // - here this is not important since we are not modifying the matrix).
+    auto matrix = mtx::create(
+        exec, gko::dim<2>(dp_3),
+        val_array::view(app_exec, (3 * dp - 2) * (3 * dp - 2) * (3 * dp - 2),
+                        values),
+        idx_array::view(app_exec, (3 * dp - 2) * (3 * dp - 2) * (3 * dp - 2),
+                        col_idxs),
+        idx_array::view(app_exec, dp_3 + 1, row_ptrs));
+
+    // RHS: similar to matrix
+    auto b = vec::create(exec, gko::dim<2>(dp_3, 1),
+                         val_array::view(app_exec, dp_3, rhs), 1);
+
+    // Solution: we have to be careful here - if the executors are different,
+    // once we compute the solution the array will not be automatically copied
+    // back to the original memory locations. Fortunately, whenever `apply` is
+    // called on a linear operator (e.g. matrix, solver) the arguments
+    // automatically get copied to the executor where the operator is, and
+    // copied back once the operation is completed. Thus, in this case, we can
+    // just define the solution on `app_exec`, and it will be automatically
+    // transferred to/from `exec` if needed.
+    auto x = vec::create(app_exec, gko::dim<2>(dp_3, 1),
+                         val_array::view(app_exec, dp_3, u), 1);
+
+    // Generate solver
+    auto solver_gen =
+        cg::build()
+            .with_criteria(
+                gko::stop::Iteration::build().with_max_iters(dp_3).on(exec),
+                gko::stop::ResidualNormReduction<>::build()
+                    .with_reduction_factor(accuracy)
+                    .on(exec))
+            .with_preconditioner(bj::build().on(exec))
+            .on(exec);
+    auto solver = solver_gen->generate(gko::give(matrix));
+
+    // Solve system
+    solver->apply(gko::lend(b), gko::lend(x));
+}
+
+int main(int argc, char *argv[])
+{
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " DISCRETIZATION_POINTS [executor]"
```

Comment:
You also need to update the usage of the function here since you added `alpha_c`, ...

Reply:
I'm not too sure if I expanded this one correctly